src/cryptonote_core/blockchain.cpp

```diff
@@ -3682,7 +3684,7 @@ bool Blockchain::is_tx_spendtime_unlocked(uint64_t unlock_time) const
   else
   {
     //interpret as time
-    uint64_t current_time = static_cast<uint64_t>(time(NULL));
+    const uint64_t current_time = hf_version >= 13 ? get_adjusted_time(m_db->height()) : time(NULL);
```

Comment:
Can't recall. It seems pointless anyway.
The whole point is to start using blockchain timestamps instead of local time. This is intended. A 30 minute delay is fine, since this is typically used to lock things long term, not for a few minutes I think ? Define broken. Adding an offset would need changing some tests IIRC, and means you might unlck too early instead of too late. Why do you think this may be needed ?

Reply:
"A 30 minute delay is fine, since this is typically used to lock things long term, not for a few minutes I think ?" Presumably if someone is locking a transaction with UTC time, they are trying to be more precise than a block height lock could provide. As the PR stands, users will have to wait an extra 60mins on average for their outputs to be unlocked (relative to the unlock time).

It sounds like you are concerned about future times being used for timelock checks. What about correcting the offset by doing a minimum of the current block's block time and the median block's time + 60mins? If blocks are accelerating then the current block's time will be more accurate than the median + 60, and if the current block has a future time then the median + 60 should be closer to the real time.

Note that even without the offset someone who can manipulate the median could push it into the future up to +1hr (since block times are limited by `local_time + 2hr` according to `check_block_timestamp()`. Adding the corrected offset allows timelocked tx to be spent up to 2hrs early in the worst case, while being much closer to +/- 10mins in the typical case.

Reply:
The concern goes in both directions, faster and slower blocks, meaning both block timestamps into the future and the past. I'd argue a window of +-60 minutes is the time precission that we can provide and that it is good enough. If you lock a transaction's outputs with unix time exactly a year into the future, you get a guaranteed precision of +-60 minutes, while if you use blocks, the precision is unknown and could vary days.

Having the ability to push the median significantly would probably be at the limit of our security boundaries, since it is miners that are pushing it. I have not done any statistics on this though.

Reply:
Sorry, rather than 'current block' for corrected offset it should be 'previous block + 2mins' since the current block may be under construction and we don't want to use local time.

Well, +/- 60mins might be ok, but I'm wondering why not push for even more precision if it's possible?

Reply:
I don't really mind either way. Let's see if other people have an opinion and I'll go with what seems the majority opinion.

Reply:
Agree with @UkoeHB, IMO it would be okay to return `the median time + 60mins` (i.e. `+ BLOCKCHAIN_TIMESTAMP_CHECK_WINDOW / 2 * DIFFICULTY_TARGET_V2` seconds).

Reply:
I'm fine with either; leaving it up to the tx creator to add an extra 60 minutes, or adding it straight into the calculation. Support seems to be in favour of adding the extra 60 minutes, so lets add it in then. It does seem more intuitive for the tx creator. 
@xiphon : What about koe's proposal to take the minimum between the last block timestamp and the median time? I have thought about it a bit and did not come up with anything that is obviously wrong or broken yet.

Reply:
@TheCharlatan

> doing a minimum of the current block's block time and the median block's time + 60mins
`The current block's block time` can't be less than `the median block's time`. Taking the minimum won't have any effect.

Reply:
It does have an effect if you compare it to the median block time + 60 minutes (which is what koe is proposing).

Reply:
Yep, i misread this. Anyway `the current block's block time` can be manipulated as well, so IMO taking minimum doesn't fix the issue.

Reply:
I feel similarly, so would suggest to go with just adding the extra 60 minutes to the `get_adjusted_time` calculation.

Reply:
Taking the minimum is meant to provide better accuracy when blocks are accelerated (appearing faster than 2mins on average). If we don't take the minimum then even if all miners are honest the 'adjusted time' will be 'in the future' and locked transactions can be unlocked early (by probably only a few minutes, but still it will be a common misalignment).

Of course, the converse will be just as common, where blocks are slowing down and the adjusted time is 'in the past'. However, it is safe for locked transaction to be unlocked late. We can think about the unlock time as 'locked until _at least_ the lock time', which is harmonious with the reality of blockchains where tx authors have to wait for miners to add txs to a block (lock time of '0' means 'as soon as possible').

What about manipulation? There are a few combinations ('adjusted time' is 'median + 60', 'current time' is block timestamp, 'real time' is e.g. UTC):
1. Current time honest
a. Median low: adjusted time < current time = real time; tx unlocked up to *chain age* late or until median improves
b. Median high: adjusted time > current time = real time; [minimum: tx unlocked on time] [no minimum: tx unlocked up to 1hr early since current time >= median time is a protocol rule]
c. Median accurate: no problem, median + 60 ~= current time, ignoring local time variances
2. Current time high (capped at ~local time + 2hr by consensus)
a. Median low: adjusted time < real time < current time; tx unlocked up to *chain age* late or until median improves; current time inaccuracy has no effect
b. Median high: adjusted time & current time > real time; tx unlocked early [minimum: tx unlocked closest to real time possible] [no minimum: tx unlocked based on adjusted time, which may be earlier than current time]
c. Median accurate: adjusted time ~= real time < current time; no problem, tx unlocked on time
3. Current time low
a. Median low: adjusted time & current time < real time; [minimum: tx unlocked as late as possible] [no minimum: tx unlocked late]
b. Median high: current time < real time < adjusted time; [minimum: tx unlocked up to 1hr late] [no minimum: tx unlocked up to 1hr early]
c. Median accurate: current time < adjusted time ~= real time; [minimum: tx unlocked up to 1hr late] [no minimum: tx unlocked on time]

Using a minimum is a safety improvement for (1b), (2b), (3b), and an inconvenience loss for (3a), (3c). In net, an improvement to the adjusted time's safety profile.

Reply:
I spent a bit of time going through some of the potential problems with koe's proposal and think it is safe enough to go with it. It does introduce some extra complexity though and detecting a few of the corner cases has been challenging. The proposal does allow a miner to block inclusion of a time locked transaction into the next block by setting a low timestamp of his own mined block. In the current protocol, there are no incentives or rewards gained by doing this. Given this added complexity I would like to get a bit more review here.

Reply:
To get the median I think it should be like this:
```
// project the median to match approximately when the block being validated will appear
// the median is calculated from a chunk of past blocks, so we use +1 to offset onto the current block
median_ts += (BLOCKCHAIN_TIMESTAMP_CHECK_WINDOW + 1) * DIFFICULTY_TARGET_V2 / 2;

// project the current block's time based on the previous block's time
// we don't use the current block's time directly to mitigate timestamp manipulation
uint64_t adjusted_current_block_ts = m_db->get_block_timestamp(height - 1) + DIFFICULTY_TARGET_V2;

// return minimum of ~current block time and adjusted median time
// we do this since it's better to report a time in the past than a time in the future
return (previous_block_ts < median_ts ? previous_block_ts : median_ts);
```
Note: on the odd looking +1 in `median_ts`... imagine:
1. window = 2, blocks are indexed b0, b1, b2; block b2 is being validated; median of window is at timestamp (b0 + b1)/2 = 0.5b; adding offset of window/2 means adjusted timestamp would be 1.5b so we need to add another 0.5b -> 0.5b + (window + 1)/2 = 2b
2. window = 3, blocks are indexed b0, b1, b2, b3; block b3 is being validated; median of window is at timestamp 1b; adding offset of window/2 means adjusted timestamp would be 2.5b so we need to add another 0.5b -> 1b + (window + 1)/2 = 3b

Reply:
I assume adjusted_current_block_ts and previous_block_ts are the same variable here, right ?

Reply:
Yeah fixed

Reply:
Pushed with that change.